### PR TITLE
Extract `Nri.Button`

### PR DESCRIPTION
**Pull request by jwoudenberg**
This extracts `Nri.Button` from the monolith into CCS.

The repo already contained a `Nri.Ui.Button.V1`, which containes a subset of an old version of `Nri.Button`. The existence of that module makes this effort `V2`. This time the plan is to immediately pull these changes back into the monolith and make `noredink-ui` the one source for button truth.

##### diff

The diff below shows a bunch of changed styles. This is just an indentation change due to the requirement to pull in assets used in these styles as an argument. The styles themselves weren't changed in this PR.

<details>
<summary>Diff of the monolith's `Nri.Button` with this PR's `Nri.Ui.Button.V2`</summary>

```diff
1c1
< module Nri.Button
---
> module Nri.Ui.Button.V2
64d63
< import AssetPath
72,74c71
< import Nri.Assets
< import Nri.Icon exposing (IconType, decorativeIcon, icon)
< import Nri.Styles
---
> import Nri.Ui.AssetPath as AssetPath exposing (Asset)
78c75,76
< import Url exposing (Url)
---
> import Nri.Ui.Icon.V2 as Icon exposing (IconType, decorativeIcon, icon)
> import Nri.Ui.Styles.V1 as Styles
288,289c286,287
< copyToClipboard : CopyToClipboardConfig -> Html msg
< copyToClipboard config =
---
> copyToClipboard : { r | teach_assignments_copyWhite_svg : Asset } -> CopyToClipboardConfig -> Html msg
> copyToClipboard assets config =
293c291
<                 Just Nri.Icon.Copy
---
>                 Just (Icon.copy assets)
414c412
<     , url : Url
---
>     , url : String
457c455
<         , url = Url.fromString (toUrl config.route)
---
>         , url = toUrl config.route
480c478
< This should only take in messages that result in tracking events. For buttons that trigger other effects on the page, please use Nri.Button.button instead
---
> This should only take in messages that result in tracking events. For buttons that trigger other effects on the page, please use Nri.Ui.Button.V2.button instead
511c509
<             :: Url.href config.url
---
>             :: Html.Attributes.href config.url
623c621
< {-| Required CSS styles for `Nri.Button`.
---
> {-| Required CSS styles for `Nri.Ui.Button.V2`.
625c623
< styles : Nri.Styles.Styles Never CssClasses msg
---
> styles : Styles.StylesWithAssets Never CssClasses msg { r | icons_xBlue_svg : Asset }
725,738c723,744
<     Nri.Styles.styles "Nri-Button-"
<         [ Css.Foreign.class Button
<             [ cursor pointer
<             , display inlineBlock
<             , -- Specifying the font can and should go away after bootstrap is removed from application.css
<               Nri.Ui.Fonts.V1.baseFont
<             , textOverflow ellipsis
<             , overflow Css.hidden
<             , textDecoration none
<             , Css.property "background-image" "none"
<             , textShadow none
<             , Css.property "transition" "all 0.2s"
<             , Css.hover
<                 [ textDecoration none
---
>     Styles.stylesWithAssets "Nri-Ui-Button-V2-" <|
>         \assets ->
>             [ Css.Foreign.class Button
>                 [ cursor pointer
>                 , display inlineBlock
>                 , -- Specifying the font can and should go away after bootstrap is removed from application.css
>                   Nri.Ui.Fonts.V1.baseFont
>                 , textOverflow ellipsis
>                 , overflow Css.hidden
>                 , textDecoration none
>                 , Css.property "background-image" "none"
>                 , textShadow none
>                 , Css.property "transition" "all 0.2s"
>                 , Css.hover
>                     [ textDecoration none
>                     ]
>                 , Css.disabled
>                     [ cursor notAllowed
>                     ]
>                 , Css.Foreign.withClass IsLink
>                     [ whiteSpace noWrap
>                     ]
740,741c746,809
<             , Css.disabled
<                 [ cursor notAllowed
---
>             , sizeStyle TinyDeprecated
>                 { fontSize = 13
>                 , height = 25
>                 , lineHeight = 25
>                 , sidePadding = 8
>                 , minWidth = 50
>                 , imageHeight = 0
>                 , shadowHeight = 0
>                 }
>             , sizeStyle Small
>                 { fontSize = 15
>                 , height = 36
>                 , lineHeight = 15
>                 , sidePadding = 16
>                 , imageHeight = 15
>                 , shadowHeight = 2
>                 , minWidth = 75
>                 }
>             , sizeStyle Medium
>                 { fontSize = 17
>                 , height = 45
>                 , lineHeight = 19
>                 , sidePadding = 16
>                 , imageHeight = 15
>                 , shadowHeight = 3
>                 , minWidth = 100
>                 }
>             , sizeStyle Large
>                 { fontSize = 20
>                 , height = 56
>                 , lineHeight = 22
>                 , sidePadding = 16
>                 , imageHeight = 20
>                 , shadowHeight = 4
>                 , minWidth = 200
>                 }
>             , styleStyle PrimaryColors
>                 { background = Nri.Ui.Colors.V1.azure
>                 , hover = Nri.Ui.Colors.V1.azureDark
>                 , text = Nri.Ui.Colors.V1.white
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.azureDark
>                 }
>             , styleStyle SecondaryColors
>                 { background = Nri.Ui.Colors.V1.white
>                 , hover = Nri.Ui.Colors.V1.glacier
>                 , text = Nri.Ui.Colors.V1.azure
>                 , border = Just <| Nri.Ui.Colors.V1.azure
>                 , shadow = Nri.Ui.Colors.V1.azure
>                 }
>             , styleStyle BorderlessColors
>                 { background = transparent
>                 , hover = transparent
>                 , text = Nri.Ui.Colors.V1.azure
>                 , border = Nothing
>                 , shadow = transparent
>                 }
>             , Css.Foreign.class (ColorsStyle BorderlessColors)
>                 [ Css.hover
>                     [ textDecoration underline
>                     , Css.disabled
>                         [ textDecoration none
>                         ]
>                     ]
743,744c811,856
<             , Css.Foreign.withClass IsLink
<                 [ whiteSpace noWrap
---
>             , styleStyle DangerColors
>                 { background = Nri.Ui.Colors.V1.red
>                 , hover = Nri.Ui.Colors.V1.redDark
>                 , text = Nri.Ui.Colors.V1.white
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.redDark
>                 }
>             , styleStyle PremiumColors
>                 { background = Nri.Ui.Colors.V1.yellow
>                 , hover = Nri.Ui.Colors.V1.ochre
>                 , text = Nri.Ui.Colors.V1.navy
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.ochre
>                 }
>             , styleStyle LoadingColors
>                 { background = Nri.Ui.Colors.V1.glacier
>                 , hover = Nri.Ui.Colors.V1.glacier
>                 , text = Nri.Ui.Colors.V1.navy
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.glacier
>                 }
>             , styleStyle ErrorColors
>                 { background = Nri.Ui.Colors.V1.purple
>                 , hover = Nri.Ui.Colors.V1.purple
>                 , text = Nri.Ui.Colors.V1.white
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.purple
>                 }
>             , styleStyle InactiveColors
>                 { background = Nri.Ui.Colors.V1.gray92
>                 , hover = Nri.Ui.Colors.V1.gray92
>                 , text = Nri.Ui.Colors.V1.gray45
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.gray92
>                 }
>             , styleStyle ActiveColors
>                 { background = Nri.Ui.Colors.V1.glacier
>                 , hover = Nri.Ui.Colors.V1.glacier
>                 , text = Nri.Ui.Colors.V1.navy
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.gray92
>                 }
>             , Css.Foreign.class (ColorsStyle ActiveColors)
>                 [ Css.boxShadow none
>                 , Css.boxShadow5 inset zero (px 3) zero (withAlpha 0.2 Nri.Ui.Colors.V1.gray20)
>                 , border3 (px 1) solid Nri.Ui.Colors.V1.azure
746,809c858,907
<             ]
<         , sizeStyle TinyDeprecated
<             { fontSize = 13
<             , height = 25
<             , lineHeight = 25
<             , sidePadding = 8
<             , minWidth = 50
<             , imageHeight = 0
<             , shadowHeight = 0
<             }
<         , sizeStyle Small
<             { fontSize = 15
<             , height = 36
<             , lineHeight = 15
<             , sidePadding = 16
<             , imageHeight = 15
<             , shadowHeight = 2
<             , minWidth = 75
<             }
<         , sizeStyle Medium
<             { fontSize = 17
<             , height = 45
<             , lineHeight = 19
<             , sidePadding = 16
<             , imageHeight = 15
<             , shadowHeight = 3
<             , minWidth = 100
<             }
<         , sizeStyle Large
<             { fontSize = 20
<             , height = 56
<             , lineHeight = 22
<             , sidePadding = 16
<             , imageHeight = 20
<             , shadowHeight = 4
<             , minWidth = 200
<             }
<         , styleStyle PrimaryColors
<             { background = Nri.Ui.Colors.V1.azure
<             , hover = Nri.Ui.Colors.V1.azureDark
<             , text = Nri.Ui.Colors.V1.white
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.azureDark
<             }
<         , styleStyle SecondaryColors
<             { background = Nri.Ui.Colors.V1.white
<             , hover = Nri.Ui.Colors.V1.glacier
<             , text = Nri.Ui.Colors.V1.azure
<             , border = Just <| Nri.Ui.Colors.V1.azure
<             , shadow = Nri.Ui.Colors.V1.azure
<             }
<         , styleStyle BorderlessColors
<             { background = transparent
<             , hover = transparent
<             , text = Nri.Ui.Colors.V1.azure
<             , border = Nothing
<             , shadow = transparent
<             }
<         , Css.Foreign.class (ColorsStyle BorderlessColors)
<             [ Css.hover
<                 [ textDecoration underline
<                 , Css.disabled
<                     [ textDecoration none
<                     ]
---
>             , styleStyle SuccessColors
>                 { background = Nri.Ui.Colors.V1.greenDark
>                 , hover = Nri.Ui.Colors.V1.greenDark
>                 , text = Nri.Ui.Colors.V1.white
>                 , border = Nothing
>                 , shadow = Nri.Ui.Colors.V1.greenDark
>                 }
>             , Css.Foreign.class (ColorsStyle PremiumColors)
>                 [ Css.property "box-shadow" "none"
>                 , marginBottom zero
>                 ]
>             , Css.Foreign.class (ColorsStyle InactiveColors)
>                 [ Css.property "box-shadow" "none"
>                 , Css.property "border" "none"
>                 , marginBottom zero
>                 ]
>             , Css.Foreign.class (ColorsStyle SuccessColors)
>                 [ Css.property "box-shadow" "none"
>                 , Css.property "border" "none"
>                 , marginBottom zero
>                 ]
>             , Css.Foreign.class (ColorsStyle ErrorColors)
>                 [ Css.property "box-shadow" "none"
>                 , Css.property "border" "none"
>                 , marginBottom zero
>                 ]
>             , Css.Foreign.class (ColorsStyle LoadingColors)
>                 [ Css.property "box-shadow" "none"
>                 , Css.property "border" "none"
>                 , marginBottom zero
>                 ]
>             , Css.Foreign.class Delete
>                 [ display inlineBlock
>                 , backgroundImage (url <| AssetPath.url assets.icons_xBlue_svg)
>                 , backgroundRepeat noRepeat
>                 , backgroundColor transparent
>                 , backgroundPosition center
>                 , backgroundSize contain
>                 , Css.property "border" "none"
>                 , Css.width (px 15)
>                 , Css.height (px 15)
>                 , margin2 zero (px 6)
>                 , cursor pointer
>                 ]
>             , Css.Foreign.class Toggled
>                 [ color Nri.Ui.Colors.V1.gray20
>                 , backgroundColor Nri.Ui.Colors.V1.glacier
>                 , boxShadow5 inset zero (px 3) zero (withAlpha 0.2 Nri.Ui.Colors.V1.gray20)
>                 , border3 (px 1) solid Nri.Ui.Colors.V1.azure
>                 , fontWeight bold
812,910d909
<         , styleStyle DangerColors
<             { background = Nri.Ui.Colors.V1.red
<             , hover = Nri.Ui.Colors.V1.redDark
<             , text = Nri.Ui.Colors.V1.white
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.redDark
<             }
<         , styleStyle PremiumColors
<             { background = Nri.Ui.Colors.V1.yellow
<             , hover = Nri.Ui.Colors.V1.ochre
<             , text = Nri.Ui.Colors.V1.navy
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.ochre
<             }
<         , styleStyle LoadingColors
<             { background = Nri.Ui.Colors.V1.glacier
<             , hover = Nri.Ui.Colors.V1.glacier
<             , text = Nri.Ui.Colors.V1.navy
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.glacier
<             }
<         , styleStyle ErrorColors
<             { background = Nri.Ui.Colors.V1.purple
<             , hover = Nri.Ui.Colors.V1.purple
<             , text = Nri.Ui.Colors.V1.white
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.purple
<             }
<         , styleStyle InactiveColors
<             { background = Nri.Ui.Colors.V1.gray92
<             , hover = Nri.Ui.Colors.V1.gray92
<             , text = Nri.Ui.Colors.V1.gray45
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.gray92
<             }
<         , styleStyle ActiveColors
<             { background = Nri.Ui.Colors.V1.glacier
<             , hover = Nri.Ui.Colors.V1.glacier
<             , text = Nri.Ui.Colors.V1.navy
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.gray92
<             }
<         , Css.Foreign.class (ColorsStyle ActiveColors)
<             [ Css.boxShadow none
<             , Css.boxShadow5 inset zero (px 3) zero (withAlpha 0.2 Nri.Ui.Colors.V1.gray20)
<             , border3 (px 1) solid Nri.Ui.Colors.V1.azure
<             ]
<         , styleStyle SuccessColors
<             { background = Nri.Ui.Colors.V1.greenDark
<             , hover = Nri.Ui.Colors.V1.greenDark
<             , text = Nri.Ui.Colors.V1.white
<             , border = Nothing
<             , shadow = Nri.Ui.Colors.V1.greenDark
<             }
<         , Css.Foreign.class (ColorsStyle PremiumColors)
<             [ Css.property "box-shadow" "none"
<             , marginBottom zero
<             ]
<         , Css.Foreign.class (ColorsStyle InactiveColors)
<             [ Css.property "box-shadow" "none"
<             , Css.property "border" "none"
<             , marginBottom zero
<             ]
<         , Css.Foreign.class (ColorsStyle SuccessColors)
<             [ Css.property "box-shadow" "none"
<             , Css.property "border" "none"
<             , marginBottom zero
<             ]
<         , Css.Foreign.class (ColorsStyle ErrorColors)
<             [ Css.property "box-shadow" "none"
<             , Css.property "border" "none"
<             , marginBottom zero
<             ]
<         , Css.Foreign.class (ColorsStyle LoadingColors)
<             [ Css.property "box-shadow" "none"
<             , Css.property "border" "none"
<             , marginBottom zero
<             ]
<         , Css.Foreign.class Delete
<             [ display inlineBlock
<             , backgroundImage (url <| AssetPath.url Nri.Assets.icons_xBlue_svg)
<             , backgroundRepeat noRepeat
<             , backgroundColor transparent
<             , backgroundPosition center
<             , backgroundSize contain
<             , Css.property "border" "none"
<             , Css.width (px 15)
<             , Css.height (px 15)
<             , margin2 zero (px 6)
<             , cursor pointer
<             ]
<         , Css.Foreign.class Toggled
<             [ color Nri.Ui.Colors.V1.gray20
<             , backgroundColor Nri.Ui.Colors.V1.glacier
<             , boxShadow5 inset zero (px 3) zero (withAlpha 0.2 Nri.Ui.Colors.V1.gray20)
<             , border3 (px 1) solid Nri.Ui.Colors.V1.azure
<             , fontWeight bold
<             ]
<         ]
```

</details>

##### For review

Followed the "Moving a widget to `noredink-ui`" step of our [process in Paper][].

[process in paper]: https://paper.dropbox.com/doc/noredink-ui-widget-library-fUK6yq32g187lxw2A60a8#:uid=348369002247633329910446&h2=Moving-a-widget-to-noredink-ui

### CODE REVIEW

**Comment by brookeangel** [comment]
Hi @jwoudenberg - excited to see `Nri.Button` joining the other doodads! I won't have time to review this today, but I'm happy to review this tomorrow if it isn't urgent! (If it is urgent, feel free to reassign)

File: src/EventExtras.elm [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+{-| This is necessary to use in single-page apps (SPA) when intercepting the
+`onClick` of `<a>` tags that trigger navigation within the app,
+as a normal `onClickPreventDefault` will prevent the user from opening the link
+in a new tab/window.
+
+(From <https://github.com/elm-lang/html/issues/110>)
+
+-}
+onClickPreventDefaultForLinkWithHref : msg -> Html.Attribute msg
+onClickPreventDefaultForLinkWithHref msg =
```
Comment (brookeangel) [inline src/EventExtras.elm]: 👍 sweet! Nice that we've extracted this

File: styleguide-app/Examples/Button.elm [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+{-| -}
+type ButtonType
+    = Button
+    | Link
+    | CopyToClipboard
+
+
+{-| -}
+example :
+    { r | teach_assignments_copyWhite_svg : Asset }
```
Comment (brookeangel) [inline styleguide-app/Examples/Button.elm]: ❓ Not sure whether I'm missing a step in the build process, but the icons do not appear in the style guide for me

Comment (jwoudenberg) [reply styleguide-app/Examples/Button.elm]: Yeah, we don't have any icons in the noredink-ui repo at the moment. We should fix that, I'll create a separate PR for that.

**Comment by jwoudenberg** [comment]
Thanks for the review Brooke! I have added an item on my todolist to add some assets into this package for use in the styleguide. Going to merge this without for now in the interest of getting a PR against the monolith sooner!

**Pull request merged**